Save 'runs' to exported models

ammgws opened the pull request:
Added `runs` to the exported model to save run-time when a Text model is instantiated from an existing JSON/dict. With the current markovify there isn't really much time saved by loading an exported model, but by saving `runs` the time is more than halved.

Please let me know what you think.

Benchmarking method (see [gist](https://gist.github.com/ammgws/9d2b4d278ae281d3ef282c437c0cef5e))
a) Create new model from scratch `markovify.Text(corpus, state_size=2, chain=None)`
b) Load an existing chain JSON `markovify.Text.from_chain(chain_json, corpus=corpus)`
c) Load an existing model JSON `markovify.Text.from_json(model_json)`

**Current markovify**
- a) RasPi 16.2 secs, Desktop 0.24 secs
- b) RasPi 14.1 secs, Desktop 0.22 secs
- c) RasPi 15.4 secs, Desktop 0.25 secs

**This branch**
- a) RasPi 16.3 secs, Desktop 0.23 secs
- b) RasPi 13.7 secs, Desktop 0.21 secs
- c) RasPi 06.3 secs, Desktop 0.10 secs

RasPi = Raspberry Pi 2
Desktop = Intel i5-3570K CPU @ 3.40GHz, 8GB RAM

jsvine commented:
Thanks, Juan! I like this idea in general. Two thoughts re. tweaks:

- I'd prefer if this feature were opt-in. Including the runs increases the size of the exported JSON/dict, and some users might not want to do that.

- Let's think about other variable names we could use instead of `runs`. That name is somewhat semantically ambiguous. (I should have chosen a more descriptive name when I initially wrote this library.) Maybe `sentence_lists` or `parsed_text`? Open to other ideas, too.

ammgws commented:
Regarding the size of the exported model, what I really wanted to do was
remove input_text from the export and in place of it export either
rejoined_text or runs since IIRC input_text is only used to generate the
other two and isn't referenced by any other functions in the class. That
way the file size should be the same as it is now or even smaller (in the
case of rejoined_text since it removes certain characters etc).

However that would break compatibility with models exported using older
versions so I left it going to discuss here.

On 5 Jan 2017 9:33 am, "Jeremy Singer-Vine" <notifications@github.com>
wrote:

> Thanks, Juan! I like this idea in general. Two thoughts re. tweaks:
>
>    -
>
>    I'd prefer if this feature were opt-in. Including the runs increases
>    the size of the exported JSON/dict, and some users might not want to do
>    that.
>    -
>
>    Let's think about other variable names we could use instead of runs.
>    That name is somewhat semantically ambiguous. (I should have chosen a more
>    descriptive name when I initially wrote this library.) Maybe
>    sentence_lists or parsed_text? Open to other ideas, too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jsvine/markovify/pull/50#issuecomment-270528413>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATc742VN0kv-WmmQMRliYxxckzJoiY54ks5rPDpzgaJpZM4LaYmE>
> .
>

ekt1701 commented:
Hello, I tried to test this pull request but I'm getting this error message:

`File "/usr/local/lib/python2.7/site-packages/markovify/text.py", line 40
    "runs": self.runs
SyntaxError: invalid syntax`

I simply replaced the existing text.py with this version.

ammgws commented:
Yeah you will need to regenerate the JSON file using the updated code, since it adds extra data to the exported JSON file.

ekt1701 commented:
I tried to make a new JSON file with the updated code, however, I got this error:

```
File "makej.py", line 1, in <module>
    import markovify
  File "/usr/local/lib/python2.7/site-packages/markovify/__init__.py", line 5, in <module>
    from .text import Text, NewlineText
  File "/usr/local/lib/python2.7/site-packages/markovify/text.py", line 40
    "runs": self.runs
SyntaxError: invalid syntax
```

ammgws commented:
Try now!

ekt1701 commented:
Thank you, the syntax errors are gone.  Hopefully, tomorrow, I can test the speed difference.

jsvine commented:
Thanks, @ammgws and thanks @ekt1701. Just checking: How did the speed compare?

ekt1701 commented:
Here are my results with a text file consisting of 36,166 quotes.

Running 25 times each, returning the average runtime in seconds.

Current markovify, create new model, state size 2:
3.36153388023
Example: But chemical agriculture offered new tools that are different, but they all met and became Picasso.
----------------------------------
Current markovify, existing chain JSON, state size 2:
3.43817088127
Example: I want to see that in the Ann Arbor area.
----------------------------------
Current markovify, existing model JSON, state size 2:
4.69978396416
Example: Love and kindness from the challenges and adversity.
----------------------------------

Modified markovify, create new model, state size 2:
3.41913988113
Example: I heard dogs bark more fiercely than usual, I stood against the people who you trusted?
----------------------------------
Modified markovify, existing chain JSON, state size 2:
3.2639552784
Example: I say that, of course.
----------------------------------
Modified markovify, existing model JSON, state size 2:
2.6922947216
Example: History provides neither compensation for suffering is transformed into an extraordinary adventure.
----------------------------------

Abhishaike commented:
Hi ammgws, is there a way to extend this idea of using 'runs' to the POSified method of generating markov chains? I've attached the POSified class below. 

```
class POSifiedText(markovify.Text): 
    def word_split(self, sentence):
        words = re.split(self.word_split_pattern, sentence)
        words = [ "::".join(tag) for tag in nltk.pos_tag(words) ]
        return words
    def word_join(self, words):
        sentence = " ".join(word.split("::")[0] for word in words)
        return sentence
```

ammgws commented:
All the POSified method does is (from were I can see) is override the
functions for how the corpus is processed. The idea of 'runs' is still the
same as it subclasses markovify, so I don't see why it wouldn't work. Did
you give it a go?

On 15 Feb. 2017 03:40, "Abhishaike" <notifications@github.com> wrote:

Hi ammgws, is there a way to extend this idea of using 'runs' to the
POSified method of generating markov chains? I've attached the POSified
class below.

class POSifiedText(markovify.Text):
    def word_split(self, sentence):
        words = re.split(self.word_split_pattern, sentence)
        words = [ "::".join(tag) for tag in nltk.pos_tag(words) ]
        return words
    def word_join(self, words):
        sentence = " ".join(word.split("::")[0] for word in words)
        return sentence

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/jsvine/markovify/pull/50#issuecomment-279795903>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ATc747E8BeMrMS8ibG5kXAtSd0YKKQFoks5rcfUtgaJpZM4LaYmE>
.

Abhishaike commented:
I believe you're right, I believe I was just overthinking it :) 
The following code seems to work just fine. 

```
class EditedTextClass(POSifiedText):
    def __init__(self, input_text, state_size=3, chain=None, runs=None):
        self.input_text = input_text
        self.state_size = state_size
        self.runs = runs or list(self.generate_corpus(self.input_text))
        self.rejoined_text = self.sentence_join(map(self.word_join, self.runs))
        self.chain = chain or markovify.Chain(self.runs, state_size)
    def to_dict(self):
        return {
            "input_text": self.input_text,
            "state_size": self.state_size,
            "chain": self.chain.to_json(),
            "runs": self.runs
        }
    @classmethod
    def from_dict(cls, obj):
        return cls(
            obj["input_text"],
            state_size=obj["state_size"],
            chain=markovify.Chain.from_json(obj["chain"]),
            runs=obj["runs"]
        )
```

jsvine commented:
I'm looking to incorporate these suggestions, but it seems this PR is failing a test:

```
$ nosetests
.........F
======================================================================
FAIL: test_simple (test.test_combine.MarkovifyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jeremysinger-vine/Dropbox/personal/projects/markovify/test/test_combine.py", line 18, in test_simple
    assert(combo.chain.model == text_model.chain.model)
AssertionError
```

Here's the relevant test:

```python
class MarkovifyTest(unittest.TestCase):

    def setUp(self):
        with open(os.path.join(os.path.dirname(__file__), "texts/sherlock.txt")) as f:
            self.sherlock = f.read()

    def test_simple(self):
        text_model = markovify.Text(self.sherlock)
        combo = markovify.combine([ text_model, text_model ], [ 0.5, 0.5 ])
        assert(combo.chain.model == text_model.chain.model)
```

Do you know what might be causing the failure? It shouldn't be affected by changes to model-exporting.

ammgws commented:
It's possible that I haven't adapted the combine function correctly (assuming the test logic is correct).
I have been looking into it this past week but haven't made any progress.

ammgws commented:
OK tests should pass now.

WnndGws commented:
Great! These extra few seconds may seem insignificant, but they really make all the difference in my case!
